src/AstRunner/AstParser/NikicPhpParser/CacheableFileParser.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SensioLabs\Deptrac\AstRunner\AstParser\NikicPhpParser;
+
+use PhpParser\JsonDecoder;
+
+class CacheableFileParser
+{
+    private $parser;
+    private $cache;
+    private $cacheFile;
+    private $loaded = false;
+    private $jsonDecoder;
+
+    public function __construct(FileParser $parser)
+    {
+        $this->parser = $parser;
+        $this->cache = [];
+        $this->cacheFile = getcwd().'/.deptrac.cache';
+        $this->jsonDecoder = new JsonDecoder();
+    }
+
+    public function parse(\SplFileInfo $file): array
+    {
+        $this->load();
+
+        $realPath = $file->getRealPath();
+        $hash = sha1_file($realPath);
+
+        if (isset($this->cache[$realPath]) && $hash === $this->cache[$realPath]['hash']) {
+            return $this->cache[$realPath]['ast'];
+        }
+
+        $ast = $this->parser->parse($file);
+        $this->cache[$realPath] = ['hash' => $hash, 'ast' => $ast];
+
+        $this->write();
+
+        return $ast;
+    }
+
+    private function load(): void
+    {
+        if ($this->loaded) {
+            return;
+        }
+
+        if (!file_exists($this->cacheFile) || !is_readable($this->cacheFile)) {
+            return;
+        }
+
+        $cache = json_decode(file_get_contents($this->cacheFile), true);
+
+        $this->cache = array_map(
```

Comment:
i guess this is eating a lot of memory. may we build a more efficient way how to store the cache.

Reply:
Any ideas? ;-)

We will run some benchmarks on big projects ...

Reply:
We checked it for a bigger project. Without the cache, the memory consumption was 36 MB. With cache at 40 MB. Thus, the difference is only 4 MB.
Since it differs only minimally and we talk about few MB, you do not need to complicate it here. ;-)